Annex60/Utilities/Time/CalendarTime.mo

```diff
@@ -0,0 +1,279 @@
+       Possibly your startTime is too large.");
+
+  // iterate to find the year at initialisation
+  year :=0;
+  for i in 1:size(timeStampsNewYear,1) loop
+    if unixTimeStamp < timeStampsNewYear[i] then
+      yearIndex :=i - 1;
+      year :=firstYear + i - 2;
+      break;
+    end if;
+  end for;
+
+  // iterate to find the month at initialisation
+  epochLastMonth := timeStampsNewYear[yearIndex];
+  for i in 1:12 loop
+    if (unixTimeStamp-epochLastMonth)/3600/24 <
+      (if i==2 and isLeapYear[yearIndex] then 1 + dayInMonth[i] else dayInMonth[i]) then
+      month :=i;
+      break;
+    else
+      epochLastMonth :=epochLastMonth + (if i == 2 and isLeapYear[yearIndex]
+         then 1 + dayInMonth[i] else dayInMonth[i])*3600*24;
+    end if;
+  end for;
+
+equation
+  // compute unix time step based on found offset
+  unixTimeStamp = tim + timOff;
+
+  // fixme: Both "when" constructs need to be rewritten.
+  // Now, you use "when time > t* then ..."
+  //   This is inefficient as the integrator will go over t* and then search
+  //   for the event.
+  //   If you were to write "when time >= t* then ...", then the simulator
+  //   can schedule a time event for t=t*.
+  //   You can try this with the model
+  /* 
+  model TestWhen
+  Integer i(start=0);
+
+equation 
+when time >= pre(i)*0.1 then  // Change >= to > to see that it triggers state events
+    i = pre(i) + 1;
+  end when;
+
+end TestWhen;
+*/
+
+  // update the year when passing the epoch time stamp of the next year
+  when (unixTimeStamp > timeStampsNewYear[pre(yearIndex)+1]) then
+    yearIndex=pre(yearIndex)+1;
+    assert(yearIndex<=size(timeStampsNewYear,1),
+      "Overflow of epoch buffer: timeStampsNewYear needs to be extended beyond the year "
+        + String(firstYear+size(timeStampsNewYear,1)));
+    // fixme: I don't understand why this is an overflow.
+    // did you mean to say "Index out of range for epoch buffer..."?
+    year = pre(year) + 1;
+  end when;
+
+  // update the month when passing the last day of the current month
+  when unixTimeStamp - pre(epochLastMonth) >
+      (if pre(month)==2 and isLeapYear[yearIndex]
+        then 1 + dayInMonth[pre(month)] else dayInMonth[pre(month)])*3600*24 then
+    month = if pre(month) == 12 then 1 else pre(month) + 1;
+    epochLastMonth = pre(epochLastMonth) +
+      (if pre(month)==2 and isLeapYear[yearIndex]
+        then 1 + dayInMonth[pre(month)] else dayInMonth[pre(month)])*3600*24;
+  end when;
+
+  // compute other variables that can be computed without using when() statements
+  daysSinceEpoch = integer(floor(unixTimeStamp/3600/24));
+  // fixme: I see why you want to use Real for minutes and I think this is a good design.
+  // However, I would make weekDay, day and hour also Reals. Then, we have one design principle
+  // (e.g., day, hour, minute are all Real and users need to convert to integer if they need an integer)
+  // rather than having a special case for minute
```

Comment:
Do you suggest to just change the variable to Real or also remove the floor() functions etc that ensure that the actual output is a round number?